extension-qr-code/src/org/eclipse/birt/extension/qrcode/views/QRCodeGeneralPage.java

```diff
@@ -0,0 +1,243 @@
+			contentpane = (Composite) children[children.length - 1];
+
+			GridLayout layout = new GridLayout(3, false);
+			layout.marginLeft = 8;
+			layout.verticalSpacing = 12;
+			contentpane.setLayout(layout);
+
+			toolkit.createLabel(contentpane, "Text Content:"); //$NON-NLS-1$
+			txtText = toolkit.createText(contentpane, ""); //$NON-NLS-1$
+			GridData gd = new GridData();
+			gd.widthHint = 200;
+
+			txtText.setLayoutData(gd);
+			txtText.addFocusListener(new FocusAdapter() {
+
+				@Override
+				public void focusLost(org.eclipse.swt.events.FocusEvent e) {
+					updateModel(QRCodeItem.TEXT_PROP);
+				}
+			});
+
+			Button btnExp = toolkit.createButton(contentpane, "...", SWT.PUSH); //$NON-NLS-1$
+			btnExp.setToolTipText("Invoke Expression Builder"); //$NON-NLS-1$
+			btnExp.addSelectionListener(new SelectionAdapter() {
+
+				@Override
+				public void widgetSelected(SelectionEvent e) {
+					openExpression(txtText);
+				}
+			});
+
+			toolkit.createLabel(contentpane, "Width (dots):");
+			spDotsWidth = new Spinner(contentpane, SWT.None);
+			spDotsWidth.setMinimum(21);
+			spDotsWidth.setMaximum(2000);
+			spDotsWidth.setDigits(0);
+
+			gd = new GridData();
+			gd.widthHint = 200;
+			gd.horizontalSpan = 2;
+			spDotsWidth.setLayoutData(gd);
+			spDotsWidth.addFocusListener(new FocusAdapter() {
+
+				@Override
+				public void focusLost(org.eclipse.swt.events.FocusEvent e) {
+					updateModel(QRCodeItem.DOTS_WIDTH_PROP);
+				}
+			});
+
+			toolkit.createLabel(contentpane, "Encoding:"); //$NON-NLS-1$
+			txtEncoding = toolkit.createText(contentpane, ""); //$NON-NLS-1$
+			gd = new GridData();
+			gd.widthHint = 200;
+			gd.horizontalSpan = 2;
+			txtEncoding.setLayoutData(gd);
+			txtEncoding.addFocusListener(new FocusAdapter() {
+
+				@Override
+				public void focusLost(org.eclipse.swt.events.FocusEvent e) {
+					updateModel(QRCodeItem.ENCODING_PROP);
+				}
+			});
+
+		}
+	}
+
+	private void openExpression(Text textControl) {
+		QRCodeItem item = getItem();
+
+		if (item != null) {
+			String oldValue = textControl.getText();
+
+			ExpressionBuilder eb = new ExpressionBuilder(textControl.getShell(), oldValue);
+			eb.setExpressionProvider(new ExpressionProvider(item.getModelHandle()));
+
+			String result = oldValue;
+
+			if (eb.open() == Window.OK) {
+				result = eb.getResult();
+			}
+
+			if (!oldValue.equals(result)) {
+				textControl.setText(result);
+
+				updateModel(QRCodeItem.TEXT_PROP);
+			}
+		}
+	}
+
+	@Override
+	public void setInput(Object input) {
+		this.input = input;
+	}
+
+	@Override
+	public void dispose() {
+		if (toolkit != null) {
+			toolkit.dispose();
+		}
+	}
+
+	private void adaptFormStyle(Composite comp) {
+		Control[] children = comp.getChildren();
+		for (int i = 0; i < children.length; i++) {
+			if (children[i] instanceof Composite) {
+				adaptFormStyle((Composite) children[i]);
+			}
+		}
+
+		toolkit.paintBordersFor(comp);
+		toolkit.adapt(comp);
+	}
+
+	protected QRCodeItem getItem() {
```

Comment:
It is better to return Optional<QRCodeItem> if it could be unavailable

Reply:
I don't want to be the first to use newer language features in the code base, in particular when I'm not used to them myself.